InvocationTargetException and SocketException when closing vscode with LSP enabled

One Eclipse SmartHome user reports these exceptions if he closes the Visualstudio Code IDE while having a lsp connection to the ESH backend.

We wonder if you could catch those Exceptions if the user tears down the connection anyway:

```
2017-11-25 07:43:34.892 [ERROR] [nrpc.json.ConcurrentMessageProcessor] - java.net.SocketException: Connection reset
java.lang.RuntimeException: java.net.SocketException: Connection reset
        at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:103) [106:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:84) [106:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
        at java.lang.Thread.run(Thread.java:748) [?:?]
Caused by: java.net.SocketException: Connection reset
        at java.net.SocketInputStream.read(SocketInputStream.java:210) ~[?:?]
        at java.net.SocketInputStream.read(SocketInputStream.java:141) ~[?:?]
        at java.net.SocketInputStream.read(SocketInputStream.java:224) ~[?:?]
        at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:61) ~[?:?]
        ... 6 more
```

and 

```
2017-11-25 07:52:59.763 [ERROR] [eclipse.lsp4j.jsonrpc.RemoteEndpoint] - Internal error: java.lang.reflect.InvocationTargetException
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
        at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:53) ~[106:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.request(GenericEndpoint.java:105) [106:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:203) [106:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:139) [106:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:149) [106:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:77) [106:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:84) [106:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
        at java.lang.Thread.run(Thread.java:748) [?:?]
Caused by: java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
        at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:51) ~[?:?]
        ... 11 more
Caused by: java.lang.RuntimeException: java.net.SocketException: Connection reset
        at org.eclipse.lsp4j.jsonrpc.json.StreamMessageConsumer.consume(StreamMessageConsumer.java:68) ~[?:?]
        at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.notify(RemoteEndpoint.java:95) ~[?:?]
        at org.eclipse.lsp4j.jsonrpc.services.EndpointProxy.invoke(EndpointProxy.java:62) ~[?:?]
        at com.sun.proxy.$Proxy188.publishDiagnostics(Unknown Source) ~[?:?]
        at org.eclipse.xtext.ide.server.LanguageServerImpl.publishDiagnostics(LanguageServerImpl.java:342) ~[?:?]
        at org.eclipse.xtext.ide.server.LanguageServerImpl.lambda$null$2(LanguageServerImpl.java:211) ~[?:?]
        at org.eclipse.xtext.ide.server.ProjectManager.lambda$null$3(ProjectManager.java:135) ~[?:?]
        at org.eclipse.xtext.build.IncrementalBuilder$InternalStatefulIncrementalBuilder.validate(IncrementalBuilder.java:267) ~[?:?]
        at org.eclipse.xtext.build.IncrementalBuilder$InternalStatefulIncrementalBuilder.lambda$launch$6(IncrementalBuilder.java:244) ~[?:?]
        at org.eclipse.xtext.build.ClusteringStorageAwareResourceLoader.lambda$executeClustered$1(ClusteringStorageAwareResourceLoader.java:77) ~[?:?]
        at org.eclipse.xtext.xbase.lib.internal.FunctionDelegate.apply(FunctionDelegate.java:42) ~[?:?]
        at com.google.common.collect.Lists$TransformingRandomAccessList$1.transform(Lists.java:617) ~[?:?]
        at com.google.common.collect.TransformedIterator.next(TransformedIterator.java:48) ~[?:?]
        at java.util.AbstractCollection.toArray(AbstractCollection.java:141) ~[?:?]
        at java.util.ArrayList.addAll(ArrayList.java:581) ~[?:?]
        at com.google.common.collect.Iterables.addAll(Iterables.java:352) ~[?:?]
        at org.eclipse.xtext.build.ClusteringStorageAwareResourceLoader.executeClustered(ClusteringStorageAwareResourceLoader.java:80) ~[?:?]
        at org.eclipse.xtext.build.BuildContext.executeClustered(BuildContext.java:55) ~[?:?]
        at org.eclipse.xtext.build.IncrementalBuilder$InternalStatefulIncrementalBuilder.launch(IncrementalBuilder.java:251) ~[?:?]
        at org.eclipse.xtext.build.IncrementalBuilder.build(IncrementalBuilder.java:399) ~[?:?]
        at org.eclipse.xtext.build.IncrementalBuilder.build(IncrementalBuilder.java:384) ~[?:?]
        at org.eclipse.xtext.ide.server.ProjectManager.doBuild(ProjectManager.java:115) ~[?:?]
        at org.eclipse.xtext.ide.server.ProjectManager.doInitialBuild(ProjectManager.java:107) ~[?:?]
        at org.eclipse.xtext.ide.server.BuildManager.doInitialBuild(BuildManager.java:138) ~[?:?]
        at org.eclipse.xtext.ide.server.WorkspaceManager.refreshWorkspaceConfig(WorkspaceManager.java:142) ~[?:?]
        at org.eclipse.xtext.ide.server.WorkspaceManager.initialize(WorkspaceManager.java:113) ~[?:?]
        at org.eclipse.xtext.ide.server.LanguageServerImpl.lambda$initialize$3(LanguageServerImpl.java:213) ~[?:?]
        at org.eclipse.xtext.ide.server.concurrent.RequestManager.runWrite(RequestManager.java:71) ~[?:?]
        at org.eclipse.xtext.ide.server.LanguageServerImpl.initialize(LanguageServerImpl.java:216) ~[?:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
        at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:51) ~[?:?]
        ... 11 more
Caused by: java.net.SocketException: Connection reset
        at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:115) ~[?:?]
        at java.net.SocketOutputStream.write(SocketOutputStream.java:143) ~[?:?]
        at org.eclipse.lsp4j.jsonrpc.json.StreamMessageConsumer.consume(StreamMessageConsumer.java:63) ~[?:?]
        at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.notify(RemoteEndpoint.java:95) ~[?:?]
        at org.eclipse.lsp4j.jsonrpc.services.EndpointProxy.invoke(EndpointProxy.java:62) ~[?:?]
        at com.sun.proxy.$Proxy188.publishDiagnostics(Unknown Source) ~[?:?]
        at org.eclipse.xtext.ide.server.LanguageServerImpl.publishDiagnostics(LanguageServerImpl.java:342) ~[?:?]
        at org.eclipse.xtext.ide.server.LanguageServerImpl.lambda$null$2(LanguageServerImpl.java:211) ~[?:?]
        at org.eclipse.xtext.ide.server.ProjectManager.lambda$null$3(ProjectManager.java:135) ~[?:?]
        at org.eclipse.xtext.build.IncrementalBuilder$InternalStatefulIncrementalBuilder.validate(IncrementalBuilder.java:267) ~[?:?]
        at org.eclipse.xtext.build.IncrementalBuilder$InternalStatefulIncrementalBuilder.lambda$launch$6(IncrementalBuilder.java:244) ~[?:?]
        at org.eclipse.xtext.build.ClusteringStorageAwareResourceLoader.lambda$executeClustered$1(ClusteringStorageAwareResourceLoader.java:77) ~[?:?]
        at org.eclipse.xtext.xbase.lib.internal.FunctionDelegate.apply(FunctionDelegate.java:42) ~[?:?]
        at com.google.common.collect.Lists$TransformingRandomAccessList$1.transform(Lists.java:617) ~[?:?]
        at com.google.common.collect.TransformedIterator.next(TransformedIterator.java:48) ~[?:?]
        at java.util.AbstractCollection.toArray(AbstractCollection.java:141) ~[?:?]
        at java.util.ArrayList.addAll(ArrayList.java:581) ~[?:?]
        at com.google.common.collect.Iterables.addAll(Iterables.java:352) ~[?:?]
        at org.eclipse.xtext.build.ClusteringStorageAwareResourceLoader.executeClustered(ClusteringStorageAwareResourceLoader.java:80) ~[?:?]
        at org.eclipse.xtext.build.BuildContext.executeClustered(BuildContext.java:55) ~[?:?]
        at org.eclipse.xtext.build.IncrementalBuilder$InternalStatefulIncrementalBuilder.launch(IncrementalBuilder.java:251) ~[?:?]
        at org.eclipse.xtext.build.IncrementalBuilder.build(IncrementalBuilder.java:399) ~[?:?]
        at org.eclipse.xtext.build.IncrementalBuilder.build(IncrementalBuilder.java:384) ~[?:?]
        at org.eclipse.xtext.ide.server.ProjectManager.doBuild(ProjectManager.java:115) ~[?:?]
        at org.eclipse.xtext.ide.server.ProjectManager.doInitialBuild(ProjectManager.java:107) ~[?:?]
        at org.eclipse.xtext.ide.server.BuildManager.doInitialBuild(BuildManager.java:138) ~[?:?]
        at org.eclipse.xtext.ide.server.WorkspaceManager.refreshWorkspaceConfig(WorkspaceManager.java:142) ~[?:?]
        at org.eclipse.xtext.ide.server.WorkspaceManager.initialize(WorkspaceManager.java:113) ~[?:?]
        at org.eclipse.xtext.ide.server.LanguageServerImpl.lambda$initialize$3(LanguageServerImpl.java:213) ~[?:?]
        at org.eclipse.xtext.ide.server.concurrent.RequestManager.runWrite(RequestManager.java:71) ~[?:?]
        at org.eclipse.xtext.ide.server.LanguageServerImpl.initialize(LanguageServerImpl.java:216) ~[?:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[?:?]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[?:?]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:?]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[?:?]
        at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:51) ~[?:?]
        ... 11 more
2017-11-25 07:52:59.824 [ERROR] [nrpc.json.ConcurrentMessageProcessor] - java.net.SocketException: Broken pipe (Write failed)
java.lang.RuntimeException: java.net.SocketException: Broken pipe (Write failed)
        at org.eclipse.lsp4j.jsonrpc.json.StreamMessageConsumer.consume(StreamMessageConsumer.java:68) [106:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleRequest(RemoteEndpoint.java:208) [106:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:139) [106:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:149) [106:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:77) [106:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:84) [106:org.eclipse.lsp4j.jsonrpc:0.2.1.v20170706-0855]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:?]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:?]
        at java.lang.Thread.run(Thread.java:748) [?:?]
Caused by: java.net.SocketException: Broken pipe (Write failed)
        at java.net.SocketOutputStream.socketWrite0(Native Method) ~[?:?]
        at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:111) ~[?:?]
        at java.net.SocketOutputStream.write(SocketOutputStream.java:143) ~[?:?]
        at org.eclipse.lsp4j.jsonrpc.json.StreamMessageConsumer.consume(StreamMessageConsumer.java:63) ~[?:?]
        ... 10 more
```

The issue can be found here: https://github.com/eclipse/smarthome/issues/4629

--- a/org.eclipse.lsp4j.generator/src/main/xtend-gen/org/eclipse/lsp4j/generator/JsonRpcDataTransformationContext.java
+++ b/org.eclipse.lsp4j.generator/src/main/xtend-gen/org/eclipse/lsp4j/generator/JsonRpcDataTransformationContext.java
@@ -167,108 +167,76 @@ public class JsonRpcDataTransformationContext implements TransformationContext {
     throw new IllegalStateException(("Unexpected type reference: " + type));
   }
   
-  public boolean isSource(final Element arg0) {
-    return this.delegate.isSource(arg0);
+  public Element getPrimaryGeneratedJavaElement(final Element source) {
+    return this.delegate.getPrimaryGeneratedJavaElement(source);
   }
   
-  public boolean isGenerated(final Element arg0) {
-    return this.delegate.isGenerated(arg0);
+  public Element getPrimarySourceElement(final Element target) {
+    return this.delegate.getPrimarySourceElement(target);
   }
   
-  public boolean isExternal(final Element arg0) {
-    return this.delegate.isExternal(arg0);
+  public boolean isExternal(final Element element) {
+    return this.delegate.isExternal(element);
   }
   
-  public Element getPrimaryGeneratedJavaElement(final Element arg0) {
-    return this.delegate.getPrimaryGeneratedJavaElement(arg0);
+  public boolean isGenerated(final Element element) {
+    return this.delegate.isGenerated(element);
   }
   
-  public boolean isThePrimaryGeneratedJavaElement(final Element arg0) {
-    return this.delegate.isThePrimaryGeneratedJavaElement(arg0);
+  public boolean isSource(final Element element) {
+    return this.delegate.isSource(element);
   }
   
-  public Element getPrimarySourceElement(final Element arg0) {
-    return this.delegate.getPrimarySourceElement(arg0);
+  public boolean isThePrimaryGeneratedJavaElement(final Element target) {
+    return this.delegate.isThePrimaryGeneratedJavaElement(target);
   }
   
-  public List<? extends Problem> getProblems(final Element arg0) {
-    return this.delegate.getProblems(arg0);
+  public void addError(final Element element, final String message) {
+    this.delegate.addError(element, message);
   }
   
-  public void addError(final Element arg0, final String arg1) {
-    this.delegate.addError(arg0, arg1);
+  public void addWarning(final Element element, final String message) {
+    this.delegate.addWarning(element, message);
   }
   
-  public void addWarning(final Element arg0, final String arg1) {
-    this.delegate.addWarning(arg0, arg1);
+  public List<? extends Problem> getProblems(final Element element) {
+    return this.delegate.getProblems(element);
   }
   
-  public void validateLater(final Procedure0 arg0) {
-    this.delegate.validateLater(arg0);
-  }
-  
-  public TypeReference newArrayTypeReference(final TypeReference arg0) {
-    return this.delegate.newArrayTypeReference(arg0);
-  }
-  
-  public TypeReference newTypeReference(final String arg0, final TypeReference... arg1) {
-    return this.delegate.newTypeReference(arg0, arg1);
-  }
-  
-  public TypeReference newTypeReference(final Type arg0, final TypeReference... arg1) {
-    return this.delegate.newTypeReference(arg0, arg1);
-  }
-  
-  public TypeReference newSelfTypeReference(final Type arg0) {
-    return this.delegate.newSelfTypeReference(arg0);
-  }
-  
-  public TypeReference newTypeReference(final Class<?> arg0, final TypeReference... arg1) {
-    return this.delegate.newTypeReference(arg0, arg1);
-  }
-  
-  public TypeReference newWildcardTypeReference() {
-    return this.delegate.newWildcardTypeReference();
-  }
-  
-  public TypeReference newWildcardTypeReference(final TypeReference arg0) {
-    return this.delegate.newWildcardTypeReference(arg0);
-  }
-  
-  public TypeReference newWildcardTypeReferenceWithLowerBound(final TypeReference arg0) {
-    return this.delegate.newWildcardTypeReferenceWithLowerBound(arg0);
-  }
-  
-  public TypeReference getObject() {
-    return this.delegate.getObject();
-  }
-  
-  public TypeReference getString() {
-    return this.delegate.getString();
-  }
-  
-  public TypeReference getList(final TypeReference arg0) {
-    return this.delegate.getList(arg0);
-  }
-  
-  public TypeReference getSet(final TypeReference arg0) {
-    return this.delegate.getSet(arg0);
+  public void validateLater(final Procedure0 validationCallback) {
+    this.delegate.validateLater(validationCallback);
   }
   
   public TypeReference getAnyType() {
     return this.delegate.getAnyType();
   }
   
-  public TypeReference getPrimitiveVoid() {
-    return this.delegate.getPrimitiveVoid();
+  public TypeReference getList(final TypeReference param) {
+    return this.delegate.getList(param);
+  }
+  
+  public TypeReference getObject() {
+    return this.delegate.getObject();
   }
   
   public TypeReference getPrimitiveBoolean() {
     return this.delegate.getPrimitiveBoolean();
   }
   
-  public TypeReference getPrimitiveShort() {
-    return this.delegate.getPrimitiveShort();
+  public TypeReference getPrimitiveByte() {
+    return this.delegate.getPrimitiveByte();
+  }
+  
+  public TypeReference getPrimitiveChar() {
+    return this.delegate.getPrimitiveChar();
+  }
+  
+  public TypeReference getPrimitiveDouble() {
+    return this.delegate.getPrimitiveDouble();
+  }
+  
+  public TypeReference getPrimitiveFloat() {
+    return this.delegate.getPrimitiveFloat();
   }
   
   public TypeReference getPrimitiveInt() {
@@ -279,132 +247,164 @@ public class JsonRpcDataTransformationContext implements TransformationContext {
     return this.delegate.getPrimitiveLong();
   }
   
-  public TypeReference getPrimitiveFloat() {
-    return this.delegate.getPrimitiveFloat();
+  public TypeReference getPrimitiveShort() {
+    return this.delegate.getPrimitiveShort();
   }
   
-  public TypeReference getPrimitiveDouble() {
-    return this.delegate.getPrimitiveDouble();
+  public TypeReference getPrimitiveVoid() {
+    return this.delegate.getPrimitiveVoid();
   }
   
-  public TypeReference getPrimitiveChar() {
-    return this.delegate.getPrimitiveChar();
+  public TypeReference getSet(final TypeReference param) {
+    return this.delegate.getSet(param);
   }
   
-  public TypeReference getPrimitiveByte() {
-    return this.delegate.getPrimitiveByte();
+  public TypeReference getString() {
+    return this.delegate.getString();
   }
   
-  public MutableClassDeclaration findClass(final String arg0) {
-    return this.delegate.findClass(arg0);
+  public TypeReference newArrayTypeReference(final TypeReference componentType) {
+    return this.delegate.newArrayTypeReference(componentType);
   }
   
-  public MutableInterfaceDeclaration findInterface(final String arg0) {
-    return this.delegate.findInterface(arg0);
+  public TypeReference newSelfTypeReference(final Type typeDeclaration) {
+    return this.delegate.newSelfTypeReference(typeDeclaration);
   }
   
-  public MutableEnumerationTypeDeclaration findEnumerationType(final String arg0) {
-    return this.delegate.findEnumerationType(arg0);
+  public TypeReference newTypeReference(final String typeName, final TypeReference... typeArguments) {
+    return this.delegate.newTypeReference(typeName, typeArguments);
   }
   
-  public MutableAnnotationTypeDeclaration findAnnotationType(final String arg0) {
-    return this.delegate.findAnnotationType(arg0);
+  public TypeReference newTypeReference(final Type typeDeclaration, final TypeReference... typeArguments) {
+    return this.delegate.newTypeReference(typeDeclaration, typeArguments);
   }
   
-  public Type findTypeGlobally(final Class<?> arg0) {
-    return this.delegate.findTypeGlobally(arg0);
+  public TypeReference newTypeReference(final Class<?> clazz, final TypeReference... typeArguments) {
+    return this.delegate.newTypeReference(clazz, typeArguments);
   }
   
-  public Type findTypeGlobally(final String arg0) {
-    return this.delegate.findTypeGlobally(arg0);
+  public TypeReference newWildcardTypeReference() {
+    return this.delegate.newWildcardTypeReference();
   }
   
-  public Iterable<? extends Path> getChildren(final Path arg0) {
-    return this.delegate.getChildren(arg0);
+  public TypeReference newWildcardTypeReference(final TypeReference upperBound) {
+    return this.delegate.newWildcardTypeReference(upperBound);
   }
   
-  public boolean exists(final Path arg0) {
-    return this.delegate.exists(arg0);
+  public TypeReference newWildcardTypeReferenceWithLowerBound(final TypeReference lowerBound) {
+    return this.delegate.newWildcardTypeReferenceWithLowerBound(lowerBound);
   }
   
-  public boolean isFolder(final Path arg0) {
-    return this.delegate.isFolder(arg0);
+  public MutableAnnotationTypeDeclaration findAnnotationType(final String qualifiedName) {
+    return this.delegate.findAnnotationType(qualifiedName);
   }
   
-  public boolean isFile(final Path arg0) {
-    return this.delegate.isFile(arg0);
+  public MutableClassDeclaration findClass(final String qualifiedName) {
+    return this.delegate.findClass(qualifiedName);
   }
   
-  public long getLastModification(final Path arg0) {
-    return this.delegate.getLastModification(arg0);
+  public MutableEnumerationTypeDeclaration findEnumerationType(final String qualifiedName) {
+    return this.delegate.findEnumerationType(qualifiedName);
   }
   
-  public String getCharset(final Path arg0) {
-    return this.delegate.getCharset(arg0);
+  public MutableInterfaceDeclaration findInterface(final String qualifiedName) {
+    return this.delegate.findInterface(qualifiedName);
   }
   
-  public CharSequence getContents(final Path arg0) {
-    return this.delegate.getContents(arg0);
+  public Type findTypeGlobally(final Class<?> clazz) {
+    return this.delegate.findTypeGlobally(clazz);
   }
   
-  public InputStream getContentsAsStream(final Path arg0) {
-    return this.delegate.getContentsAsStream(arg0);
+  public Type findTypeGlobally(final String typeName) {
+    return this.delegate.findTypeGlobally(typeName);
   }
   
-  public URI toURI(final Path arg0) {
-    return this.delegate.toURI(arg0);
+  public boolean exists(final Path path) {
+    return this.delegate.exists(path);
   }
   
-  public Path getSourceFolder(final Path arg0) {
-    return this.delegate.getSourceFolder(arg0);
+  public String getCharset(final Path path) {
+    return this.delegate.getCharset(path);
   }
   
-  public Path getTargetFolder(final Path arg0) {
-    return this.delegate.getTargetFolder(arg0);
+  public Iterable<? extends Path> getChildren(final Path path) {
+    return this.delegate.getChildren(path);
   }
   
-  public Path getProjectFolder(final Path arg0) {
-    return this.delegate.getProjectFolder(arg0);
+  public CharSequence getContents(final Path path) {
+    return this.delegate.getContents(path);
   }
   
-  public Set<Path> getProjectSourceFolders(final Path arg0) {
-    return this.delegate.getProjectSourceFolders(arg0);
+  public InputStream getContentsAsStream(final Path path) {
+    return this.delegate.getContentsAsStream(path);
   }
   
-  public AnnotationReference newAnnotationReference(final String arg0) {
-    return this.delegate.newAnnotationReference(arg0);
+  public long getLastModification(final Path path) {
+    return this.delegate.getLastModification(path);
   }
   
-  public AnnotationReference newAnnotationReference(final Type arg0) {
-    return this.delegate.newAnnotationReference(arg0);
+  public boolean isFile(final Path path) {
+    return this.delegate.isFile(path);
   }
   
-  public AnnotationReference newAnnotationReference(final Class<?> arg0) {
-    return this.delegate.newAnnotationReference(arg0);
+  public boolean isFolder(final Path path) {
+    return this.delegate.isFolder(path);
   }
   
-  public AnnotationReference newAnnotationReference(final AnnotationReference arg0) {
-    return this.delegate.newAnnotationReference(arg0);
+  public URI toURI(final Path path) {
+    return this.delegate.toURI(path);
   }
   
-  public AnnotationReference newAnnotationReference(final String arg0, final Procedure1<AnnotationReferenceBuildContext> arg1) {
-    return this.delegate.newAnnotationReference(arg0, arg1);
+  public Path getProjectFolder(final Path path) {
+    return this.delegate.getProjectFolder(path);
   }
   
-  public AnnotationReference newAnnotationReference(final Type arg0, final Procedure1<AnnotationReferenceBuildContext> arg1) {
-    return this.delegate.newAnnotationReference(arg0, arg1);
+  public Set<Path> getProjectSourceFolders(final Path path) {
+    return this.delegate.getProjectSourceFolders(path);
   }
   
-  public AnnotationReference newAnnotationReference(final Class<?> arg0, final Procedure1<AnnotationReferenceBuildContext> arg1) {
-    return this.delegate.newAnnotationReference(arg0, arg1);
+  public Path getSourceFolder(final Path path) {
+    return this.delegate.getSourceFolder(path);
   }
   
-  public AnnotationReference newAnnotationReference(final AnnotationReference arg0, final Procedure1<AnnotationReferenceBuildContext> arg1) {
-    return this.delegate.newAnnotationReference(arg0, arg1);
+  public Path getTargetFolder(final Path sourceFolder) {
+    return this.delegate.getTargetFolder(sourceFolder);
   }
   
-  public void setPrimarySourceElement(final MutableElement arg0, final Element arg1) {
-    this.delegate.setPrimarySourceElement(arg0, arg1);
+  public AnnotationReference newAnnotationReference(final String annotationTypeName) {
+    return this.delegate.newAnnotationReference(annotationTypeName);
+  }
+  
+  public AnnotationReference newAnnotationReference(final Type annotationTypeDelcaration) {
+    return this.delegate.newAnnotationReference(annotationTypeDelcaration);
+  }
+  
+  public AnnotationReference newAnnotationReference(final Class<?> annotationClass) {
+    return this.delegate.newAnnotationReference(annotationClass);
+  }
+  
+  public AnnotationReference newAnnotationReference(final AnnotationReference annotationReference) {
+    return this.delegate.newAnnotationReference(annotationReference);
+  }
+  
+  public AnnotationReference newAnnotationReference(final String annotationTypeName, final Procedure1<AnnotationReferenceBuildContext> initializer) {
+    return this.delegate.newAnnotationReference(annotationTypeName, initializer);
+  }
+  
+  public AnnotationReference newAnnotationReference(final Type annotationTypeDelcaration, final Procedure1<AnnotationReferenceBuildContext> initializer) {
+    return this.delegate.newAnnotationReference(annotationTypeDelcaration, initializer);
+  }
+  
+  public AnnotationReference newAnnotationReference(final Class<?> annotationClass, final Procedure1<AnnotationReferenceBuildContext> initializer) {
+    return this.delegate.newAnnotationReference(annotationClass, initializer);
+  }
+  
+  public AnnotationReference newAnnotationReference(final AnnotationReference annotationReference, final Procedure1<AnnotationReferenceBuildContext> initializer) {
+    return this.delegate.newAnnotationReference(annotationReference, initializer);
+  }
+  
+  public void setPrimarySourceElement(final MutableElement javaElement, final Element sourceElement) {
+    this.delegate.setPrimarySourceElement(javaElement, sourceElement);
   }
   
   @Pure

--- a/org.eclipse.lsp4j.jsonrpc.debug/src/test/java/org/eclipse/lsp4j/jsonrpc/debug/test/LogMessageAccumulator.java
+++ b/org.eclipse.lsp4j.jsonrpc.debug/src/test/java/org/eclipse/lsp4j/jsonrpc/debug/test/LogMessageAccumulator.java
@@ -23,23 +23,24 @@ public class LogMessageAccumulator extends Handler {
 	private static final long TIMEOUT = 2000;
 
 	private final List<LogRecord> records = new ArrayList<>();
-
 	private final List<Logger> registeredLoggers = new ArrayList<>();
-
+	
 	public Logger registerTo(Class<?> clazz) {
 		return registerTo(clazz.getName());
 	}
-
+	
 	public Logger registerTo(String name) {
 		Logger logger = Logger.getLogger(name);
 		logger.setUseParentHandlers(false);
 		logger.addHandler(this);
+		logger.setLevel(Level.ALL);
 		registeredLoggers.add(logger);
 		return logger;
 	}
-
+	
 	public void unregister() {
 		for (Logger logger : registeredLoggers) {
+			logger.setLevel(null);
 			logger.removeHandler(this);
 			logger.setUseParentHandlers(true);
 		}

--- a/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/JsonRpcException.java
+++ b/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/JsonRpcException.java
@@ -1,0 +1,35 @@
+/*******************************************************************************
+ * Copyright (c) 2018 TypeFox GmbH (http://www.typefox.io) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.lsp4j.jsonrpc;
+
+import java.io.InterruptedIOException;
+import java.net.SocketException;
+import java.nio.channels.ClosedChannelException;
+
+/**
+ * An exception thrown when accessing the JSON-RPC communication channel fails.
+ */
+public class JsonRpcException extends RuntimeException {
+	
+	/**
+	 * Whether the given exception indicates that the currently accessed stream has been closed.
+	 */
+	public static boolean indicatesStreamClosed(Throwable thr) {
+		return thr instanceof InterruptedIOException
+				|| thr instanceof ClosedChannelException
+				|| thr instanceof SocketException && ("Connection reset".equals(thr.getMessage()) || "Socket closed".equals(thr.getMessage()))
+				|| thr instanceof JsonRpcException && indicatesStreamClosed(thr.getCause());
+	}
+
+	private static final long serialVersionUID = -7952794305289314670L;
+	
+	public JsonRpcException(Throwable cause) {
+		super(cause);
+	}
+
+}

--- a/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/MessageConsumer.java
+++ b/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/MessageConsumer.java
@@ -15,7 +15,8 @@ public interface MessageConsumer {
 	 * Consume a single message.
 	 * 
 	 * @throws MessageIssueException when an issue is found that prevents further processing of the message
+	 * @throws JsonRpcException when accessing the JSON-RPC communication channel fails
 	 */
-	void consume(Message message) throws MessageIssueException;
+	void consume(Message message) throws MessageIssueException, JsonRpcException;
 	
 }

--- a/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/MessageProducer.java
+++ b/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/MessageProducer.java
@@ -12,7 +12,9 @@ public interface MessageProducer {
 	/**
 	 * Listen to a message source and forward all messages to the given consumer. Typically this method
 	 * blocks until the message source is unable to deliver more messages.
+	 * 
+	 * @throws JsonRpcException when accessing the JSON-RPC communication channel fails
 	 */
-	void listen(MessageConsumer messageConsumer);
+	void listen(MessageConsumer messageConsumer) throws JsonRpcException;
 	
 }

--- a/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/RemoteEndpoint.java
+++ b/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/RemoteEndpoint.java
@@ -118,7 +118,12 @@ public class RemoteEndpoint implements Endpoint, MessageConsumer, MessageIssueHa
 	@Override
 	public void notify(String method, Object parameter) {
 		NotificationMessage notificationMessage = createNotificationMessage(method, parameter);
-		out.consume(notificationMessage);
+		try {
+			out.consume(notificationMessage);
+		} catch (Exception exception) {
+			Level logLevel = JsonRpcException.indicatesStreamClosed(exception) ? Level.INFO : Level.WARNING;
+			LOG.log(logLevel, "Failed to send notification message.", exception);
+		}
 	}
 
 	protected NotificationMessage createNotificationMessage(String method, Object parameter) {

--- a/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/StreamMessageConsumer.java
+++ b/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/StreamMessageConsumer.java
@@ -11,6 +11,7 @@ import java.io.IOException;
 import java.io.OutputStream;
 import java.nio.charset.StandardCharsets;
 
+import org.eclipse.lsp4j.jsonrpc.JsonRpcException;
 import org.eclipse.lsp4j.jsonrpc.MessageConsumer;
 import org.eclipse.lsp4j.jsonrpc.messages.Message;
 
@@ -63,8 +64,8 @@ public class StreamMessageConsumer implements MessageConsumer, MessageConstants 
 				output.write(contentBytes);
 				output.flush();
 			}
-		} catch (IOException e) {
-			throw new RuntimeException(e);
+		} catch (IOException exception) {
+			throw new JsonRpcException(exception);
 		}
 	}
 

--- a/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/StreamMessageProducer.java
+++ b/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/json/StreamMessageProducer.java
@@ -10,12 +10,11 @@ package org.eclipse.lsp4j.jsonrpc.json;
 import java.io.Closeable;
 import java.io.IOException;
 import java.io.InputStream;
-import java.io.InterruptedIOException;
-import java.nio.channels.ClosedChannelException;
 import java.nio.charset.StandardCharsets;
 import java.util.logging.Level;
 import java.util.logging.Logger;
 
+import org.eclipse.lsp4j.jsonrpc.JsonRpcException;
 import org.eclipse.lsp4j.jsonrpc.MessageConsumer;
 import org.eclipse.lsp4j.jsonrpc.MessageIssueException;
 import org.eclipse.lsp4j.jsonrpc.MessageIssueHandler;
@@ -73,51 +72,48 @@ public class StreamMessageProducer implements MessageProducer, Closeable, Messag
 			boolean newLine = false;
 			Headers headers = new Headers();
 			while (keepRunning) {
-				try {
-					int c = input.read();
-					if (c == -1) {
-						// End of input stream has been reached
-						keepRunning = false;
-					} else {
-						if (debugBuilder == null)
-							debugBuilder = new StringBuilder();
-						debugBuilder.append((char) c);
-						if (c == '\n') {
-							if (newLine) {
-								// Two consecutive newlines have been read, which signals the start of the message content
-								if (headers.contentLength < 0) {
-									fireError(new IllegalStateException("Missing header " + CONTENT_LENGTH_HEADER
-											+ " in input \"" + debugBuilder + "\""));
-								} else {
-									boolean result = handleMessage(input, headers);
-									if (!result)
-										keepRunning = false;
-									newLine = false;
-								}
-								headers = new Headers();
-								debugBuilder = null;
-							} else if (headerBuilder != null) {
-								// A single newline ends a header line
-								parseHeader(headerBuilder.toString(), headers);
-								headerBuilder = null;
+				int c = input.read();
+				if (c == -1) {
+					// End of input stream has been reached
+					keepRunning = false;
+				} else {
+					if (debugBuilder == null)
+						debugBuilder = new StringBuilder();
+					debugBuilder.append((char) c);
+					if (c == '\n') {
+						if (newLine) {
+							// Two consecutive newlines have been read, which signals the start of the message content
+							if (headers.contentLength < 0) {
+								fireError(new IllegalStateException("Missing header " + CONTENT_LENGTH_HEADER
+										+ " in input \"" + debugBuilder + "\""));
+							} else {
+								boolean result = handleMessage(input, headers);
+								if (!result)
+									keepRunning = false;
+								newLine = false;
 							}
-							newLine = true;
-						} else if (c != '\r') {
-							// Add the input to the current header line
-							if (headerBuilder == null)
-								headerBuilder = new StringBuilder();
-							headerBuilder.append((char) c);
-							newLine = false;
+							headers = new Headers();
+							debugBuilder = null;
+						} else if (headerBuilder != null) {
+							// A single newline ends a header line
+							parseHeader(headerBuilder.toString(), headers);
+							headerBuilder = null;
 						}
+						newLine = true;
+					} else if (c != '\r') {
+						// Add the input to the current header line
+						if (headerBuilder == null)
+							headerBuilder = new StringBuilder();
+						headerBuilder.append((char) c);
+						newLine = false;
 					}
-				} catch (InterruptedIOException e) {
-					// The read operation has been interrupted
-				} catch (ClosedChannelException e) {
-					// The channel whose stream has been listened was closed
 				}
 			} // while (keepRunning)
-		} catch (IOException e) {
-			throw new RuntimeException(e);
+		} catch (IOException exception) {
+			if (JsonRpcException.indicatesStreamClosed(exception))
+				fireStreamClosed(exception);
+			else
+				throw new JsonRpcException(exception);
 		} finally {
 			this.callback = null;
 			this.keepRunning = false;
@@ -128,7 +124,16 @@ public class StreamMessageProducer implements MessageProducer, Closeable, Messag
 	 * Log an error.
 	 */
 	protected void fireError(Throwable error) {
-		LOG.log(Level.SEVERE, error.getMessage(), error);
+		String message = error.getMessage() != null ? error.getMessage() : "An error occurred while processing an incoming message.";
+		LOG.log(Level.SEVERE, message, error);
+	}
+	
+	/**
+	 * Report that the stream was closed through an exception.
+	 */
+	protected void fireStreamClosed(Exception cause) {
+		String message = cause.getMessage() != null ? cause.getMessage() : "The input stream was closed.";
+		LOG.log(Level.INFO, message, cause);
 	}
 
 	/**

--- a/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/validation/ReflectiveMessageValidator.java
+++ b/org.eclipse.lsp4j.jsonrpc/src/main/java/org/eclipse/lsp4j/jsonrpc/validation/ReflectiveMessageValidator.java
@@ -19,6 +19,7 @@ import java.util.ListIterator;
 import java.util.logging.Level;
 import java.util.logging.Logger;
 
+import org.eclipse.lsp4j.jsonrpc.JsonRpcException;
 import org.eclipse.lsp4j.jsonrpc.MessageConsumer;
 import org.eclipse.lsp4j.jsonrpc.MessageIssueException;
 import org.eclipse.lsp4j.jsonrpc.messages.Message;
@@ -52,7 +53,7 @@ public class ReflectiveMessageValidator implements MessageConsumer {
 	}
 
 	@Override
-	public void consume(Message message) throws MessageIssueException {
+	public void consume(Message message) throws MessageIssueException, JsonRpcException {
 		List<MessageIssue> result = new ArrayList<>();
 		try {
 			validate(message, result, new LinkedList<>(), new LinkedList<>());

--- a/org.eclipse.lsp4j.jsonrpc/src/test/java/org/eclipse/lsp4j/jsonrpc/test/LogMessageAccumulator.java
+++ b/org.eclipse.lsp4j.jsonrpc/src/test/java/org/eclipse/lsp4j/jsonrpc/test/LogMessageAccumulator.java
@@ -24,7 +24,6 @@ public class LogMessageAccumulator extends Handler {
 	private static final long TIMEOUT = 2000;
 	
 	private final List<LogRecord> records = new ArrayList<>();
-	
 	private final List<Logger> registeredLoggers = new ArrayList<>();
 	
 	public Logger registerTo(Class<?> clazz) {
@@ -35,12 +34,14 @@ public class LogMessageAccumulator extends Handler {
 		Logger logger = Logger.getLogger(name);
 		logger.setUseParentHandlers(false);
 		logger.addHandler(this);
+		logger.setLevel(Level.ALL);
 		registeredLoggers.add(logger);
 		return logger;
 	}
 	
 	public void unregister() {
 		for (Logger logger : registeredLoggers) {
+			logger.setLevel(null);
 			logger.removeHandler(this);
 			logger.setUseParentHandlers(true);
 		}

--- a/org.eclipse.lsp4j.jsonrpc/src/test/java/org/eclipse/lsp4j/jsonrpc/test/RemoteEndpointTest.java
+++ b/org.eclipse.lsp4j.jsonrpc/src/test/java/org/eclipse/lsp4j/jsonrpc/test/RemoteEndpointTest.java
@@ -12,6 +12,7 @@ import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertTrue;
 
+import java.net.SocketException;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.LinkedHashMap;
@@ -22,8 +23,10 @@ import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.TimeUnit;
 import java.util.function.Consumer;
+import java.util.logging.Level;
 
 import org.eclipse.lsp4j.jsonrpc.Endpoint;
+import org.eclipse.lsp4j.jsonrpc.JsonRpcException;
 import org.eclipse.lsp4j.jsonrpc.MessageConsumer;
 import org.eclipse.lsp4j.jsonrpc.RemoteEndpoint;
 import org.eclipse.lsp4j.jsonrpc.messages.Either;
@@ -253,6 +256,50 @@ public class RemoteEndpointTest {
 			Assert.fail("Expected an ExecutionException.");
 		} catch (ExecutionException exception) {
 			assertEquals("java.lang.RuntimeException: BAAZ", exception.getMessage());
+		}
+	}
+	
+	@Test
+	public void testExceptionInOutputStream() throws Exception {
+		LogMessageAccumulator logMessages = new LogMessageAccumulator();
+		try {
+			logMessages.registerTo(RemoteEndpoint.class);
+			
+			TestEndpoint endp = new TestEndpoint();
+			MessageConsumer consumer = new MessageConsumer() {
+				@Override
+				public void consume(Message message) throws JsonRpcException {
+					throw new JsonRpcException(new SocketException("Permission denied: connect"));
+				}
+			};
+			RemoteEndpoint endpoint = new RemoteEndpoint(consumer, endp);
+			endpoint.notify("foo", null);
+			
+			logMessages.await(Level.WARNING, "Failed to send notification message.");
+		} finally {
+			logMessages.unregister();
+		}
+	}
+	
+	@Test
+	public void testOutputStreamClosed() throws Exception {
+		LogMessageAccumulator logMessages = new LogMessageAccumulator();
+		try {
+			logMessages.registerTo(RemoteEndpoint.class);
+			
+			TestEndpoint endp = new TestEndpoint();
+			MessageConsumer consumer = new MessageConsumer() {
+				@Override
+				public void consume(Message message) throws JsonRpcException {
+					throw new JsonRpcException(new SocketException("Socket closed"));
+				}
+			};
+			RemoteEndpoint endpoint = new RemoteEndpoint(consumer, endp);
+			endpoint.notify("foo", null);
+			
+			logMessages.await(Level.INFO, "Failed to send notification message.");
+		} finally {
+			logMessages.unregister();
 		}
 	}
 

--- a/org.eclipse.lsp4j.jsonrpc/src/test/java/org/eclipse/lsp4j/jsonrpc/test/json/MessageProducerTest.java
+++ b/org.eclipse.lsp4j.jsonrpc/src/test/java/org/eclipse/lsp4j/jsonrpc/test/json/MessageProducerTest.java
@@ -1,0 +1,120 @@
+/*******************************************************************************
+ * Copyright (c) 2018 TypeFox GmbH (http://www.typefox.io) and others.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *******************************************************************************/
+package org.eclipse.lsp4j.jsonrpc.test.json;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InterruptedIOException;
+import java.net.SocketException;
+import java.nio.channels.ClosedChannelException;
+import java.util.Collections;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import org.eclipse.lsp4j.jsonrpc.JsonRpcException;
+import org.eclipse.lsp4j.jsonrpc.json.MessageJsonHandler;
+import org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+
+public class MessageProducerTest {
+	
+	private static final long TIMEOUT = 2000;
+	
+	private ExecutorService executorService;
+	private Level logLevel;
+	
+	@Before
+	public void setup() {
+		executorService = Executors.newCachedThreadPool();
+		Logger logger = Logger.getLogger(StreamMessageProducer.class.getName());
+		logLevel = logger.getLevel();
+		logger.setLevel(Level.SEVERE);
+	}
+	
+	@After
+	public void teardown() {
+		executorService.shutdown();
+		Logger logger = Logger.getLogger(StreamMessageProducer.class.getName());
+		logger.setLevel(logLevel);
+	}
+	
+	@Test
+	public void testStopOnInterrrupt() throws Exception {
+		executorService.submit(() -> {
+			InputStream input = new InputStream() {
+				@Override
+				public int read() throws IOException {
+					throw new InterruptedIOException();
+				}
+			};
+			MessageJsonHandler jsonHandler = new MessageJsonHandler(Collections.emptyMap());
+			StreamMessageProducer messageProducer = new StreamMessageProducer(input, jsonHandler);
+			messageProducer.listen(message -> {});
+			messageProducer.close();
+		}).get(TIMEOUT, TimeUnit.MILLISECONDS);
+	}
+	
+	@Test
+	public void testStopOnChannelClosed() throws Exception {
+		executorService.submit(() -> {
+			InputStream input = new InputStream() {
+				@Override
+				public int read() throws IOException {
+					throw new ClosedChannelException();
+				}
+			};
+			MessageJsonHandler jsonHandler = new MessageJsonHandler(Collections.emptyMap());
+			StreamMessageProducer messageProducer = new StreamMessageProducer(input, jsonHandler);
+			messageProducer.listen(message -> {});
+			messageProducer.close();
+		}).get(TIMEOUT, TimeUnit.MILLISECONDS);
+	}
+	
+	@Test
+	public void testStopOnSocketClosed() throws Throwable {
+		executorService.submit(() -> {
+			InputStream input = new InputStream() {
+				@Override
+				public int read() throws IOException {
+					throw new SocketException("Socket closed");
+				}
+			};
+			MessageJsonHandler jsonHandler = new MessageJsonHandler(Collections.emptyMap());
+			StreamMessageProducer messageProducer = new StreamMessageProducer(input, jsonHandler);
+			messageProducer.listen(message -> {});
+			messageProducer.close();
+		}).get(TIMEOUT, TimeUnit.MILLISECONDS);
+	}
+	
+	@Test(expected=JsonRpcException.class)
+	public void testIOException() throws Throwable {
+		try {
+			executorService.submit(() -> {
+				InputStream input = new InputStream() {
+					@Override
+					public int read() throws IOException {
+						throw new SocketException("Permission denied: connect");
+					}
+				};
+				MessageJsonHandler jsonHandler = new MessageJsonHandler(Collections.emptyMap());
+				StreamMessageProducer messageProducer = new StreamMessageProducer(input, jsonHandler);
+				messageProducer.listen(message -> {});
+				messageProducer.close();
+			}).get(TIMEOUT, TimeUnit.MILLISECONDS);
+		} catch (ExecutionException e) {
+			throw e.getCause();
+		}
+	}
+
+}

--- a/org.eclipse.lsp4j/src/test/java/org/eclipse/lsp4j/test/LogMessageAccumulator.java
+++ b/org.eclipse.lsp4j/src/test/java/org/eclipse/lsp4j/test/LogMessageAccumulator.java
@@ -23,7 +23,6 @@ public class LogMessageAccumulator extends Handler {
 	private static final long TIMEOUT = 2000;
 	
 	private final List<LogRecord> records = new ArrayList<>();
-	
 	private final List<Logger> registeredLoggers = new ArrayList<>();
 	
 	public Logger registerTo(Class<?> clazz) {
@@ -34,12 +33,14 @@ public class LogMessageAccumulator extends Handler {
 		Logger logger = Logger.getLogger(name);
 		logger.setUseParentHandlers(false);
 		logger.addHandler(this);
+		logger.setLevel(Level.ALL);
 		registeredLoggers.add(logger);
 		return logger;
 	}
 	
 	public void unregister() {
 		for (Logger logger : registeredLoggers) {
+			logger.setLevel(null);
 			logger.removeHandler(this);
 			logger.setUseParentHandlers(true);
 		}

--- a/org.eclipse.lsp4j/src/test/java/org/eclipse/lsp4j/test/launch/LauncherTest.java
+++ b/org.eclipse.lsp4j/src/test/java/org/eclipse/lsp4j/test/launch/LauncherTest.java
@@ -18,6 +18,8 @@ import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+import java.util.logging.Logger;
 
 import org.eclipse.lsp4j.CompletionItem;
 import org.eclipse.lsp4j.CompletionItemKind;
@@ -29,6 +31,7 @@ import org.eclipse.lsp4j.Position;
 import org.eclipse.lsp4j.TextDocumentIdentifier;
 import org.eclipse.lsp4j.jsonrpc.Endpoint;
 import org.eclipse.lsp4j.jsonrpc.Launcher;
+import org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer;
 import org.eclipse.lsp4j.jsonrpc.messages.Either;
 import org.eclipse.lsp4j.jsonrpc.services.ServiceEndpoints;
 import org.eclipse.lsp4j.launch.LSPLauncher;
@@ -133,6 +136,8 @@ public class LauncherTest {
 	private Launcher<LanguageServer> clientLauncher;
 	private Future<?> clientListening;
 	
+	private Level logLevel;
+	
 	@Before public void setup() throws IOException {
 		PipedInputStream inClient = new PipedInputStream();
 		PipedOutputStream outClient = new PipedOutputStream();
@@ -148,11 +153,18 @@ public class LauncherTest {
 		client = new AssertingEndpoint();
 		clientLauncher = LSPLauncher.createClientLauncher(ServiceEndpoints.toServiceObject(client, LanguageClient.class), inClient, outClient);
 		clientListening = clientLauncher.startListening();
+		
+		Logger logger = Logger.getLogger(StreamMessageProducer.class.getName());
+		logLevel = logger.getLevel();
+		logger.setLevel(Level.SEVERE);
 	}
 	
 	@After public void teardown() throws InterruptedException, ExecutionException {
 		clientListening.cancel(true);
 		serverListening.cancel(true);
+		Thread.sleep(10);
+		Logger logger = Logger.getLogger(StreamMessageProducer.class.getName());
+		logger.setLevel(logLevel);
 	}
 
 }

